### PR TITLE
storage: Introduce AddExtraTableIDColumnTransformOp

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6827

Problem Summary:

### What is changed and how it works?

This is the remaining task of https://github.com/pingcap/tiflash/pull/7493, that excluded the AddExtraTableIDColumn logic from SegmentInputStream/SegmentSourceOp.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Settings:

```toml
[profiles.default]
dt_enable_read_thread = false
enable_pipeline = true
enforce_enable_pipeline = true
```

Query 1 (no partition):

```sql
set @@session.tidb_opt_enable_late_materialization = 'off';

/*PLACEHOLDER*/ /*q1*/select   ol_number,
         sum(ol_quantity) as sum_qty,
         sum(ol_amount) as sum_amount,
         avg(ol_quantity) as avg_qty,
         avg(ol_amount) as avg_amount,
         count(*) as count_order
from	 order_line
where	 ol_delivery_d > '2007-01-02 00:00:00.000000'
group by ol_number order by ol_number;
```

Query 2 (partitioned):

```sql
create table t1(id int, age int, key(id)) partition by range(id) (
  partition p0 values less than (100),
  partition p1 values less than (200),
  partition p2 values less than (300),
  partition p3 values less than (400));

insert into t1 values (1, 10);
insert into t1 values (10, 10);

alter table t1 set tiflash replica 1;

set @@session.tidb_isolation_read_engines='tiflash';
set @@session.tidb_partition_prune_mode = 'dynamic';

analyze table t1;

select * from information_schema.tiflash_replica;

select count(*) from t1 where id < 150;

```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Operators/AddExtraTableIDColumnTransformOp.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Logger.h>
```
Comment (JaySon-Huang) [inline dbms/src/Operators/AddExtraTableIDColumnTransformOp.h:17]:
```suggestion
```
this seems useless

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp [hunk 2]
```diff
@@ -1199,22 +1200,31 @@ void DeltaMergeStore::read(
...
+                after_segment_read,
+                columns_to_read,
+                filter,
+                max_version,
+                expected_block_size,
+                /* read_mode = */ is_fast_scan ? ReadMode::Fast : ReadMode::Normal,
+                log_tracing_id));
         }
     }
+    if (!enable_read_thread)
```
Comment (SeaRise) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:1216]: Maybe
```
if (enable_read_thread)
{
    for (size_t i = 0; i < final_num_stream; ++i)
    {
        group_builder.addConcurrency(
            std::make_unique<UnorderedSourceOp>(
                exec_status,
                read_task_pool,
                columns_to_read,
                extra_table_id_index,
                log_tracing_id));
    }
}
else
{
    for (size_t i = 0; i < final_num_stream; ++i)
    {
        group_builder.addConcurrency(std::make_unique<DMSegmentThreadSourceOp>(
            exec_status,
            dm_context,
            read_task_pool,
            after_segment_read,
            columns_to_read,
            filter,
            max_version,
            expected_block_size,
            /* read_mode = */ is_fast_scan ? ReadMode::Fast : ReadMode::Normal,
            log_tracing_id));
    }
    group_builder.transform([&](auto & builder) {
        builder.appendTransformOp(std::make_unique<AddExtraTableIDColumnTransformOp>(
            exec_status,
            log_tracing_id,
            columns_to_read,
            extra_table_id_index,
            physical_table_id));
    });
}
```
is better?

**Review by SeaRise — COMMENTED** [review]
Others LGTM

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp (hunk 2 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp]: Updated, PTAL again, thanks!

**Review by SeaRise — APPROVED** [review]

**Pull request merged**